File: Python3/401 - 600/535 Encode and Decode TinyURL.py
# 24ms, 93.05%; 11.8MB, 5.36%
class Codec:
    def __init__(self):
        self.url2tiny = {}
        self.tiny2url = []
        self.tiny_idx = -1

    def encode(self, longUrl):
        """Encodes a URL to a shortened URL.
        
        :type longUrl: str
        :rtype: str
        """
        self.tiny_idx += 1
        self.url2tiny[longUrl] = self.tiny_idx
        self.tiny2url.append(longUrl)
        return 'http://tinyurl.com/' + str(self.tiny_idx)

    def decode(self, shortUrl):
        """Decodes a shortened URL to its original URL.
        
        :type shortUrl: str
        :rtype: str
        """
        return self.tiny2url[int(shortUrl.split('/')[-1])]

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(url))

File: 401 - 600/535. Encode and Decode TinyURL.cc
// 4ms, 100.00%; 9.2MB, 55.29%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
private:
    int tinyIndex = -1;
    vector<string> tiny2Url;
    unordered_map<string, int> url2Tiny;

public:
    // Encodes a URL to a shortened URL.
    string encode(const string &longUrl) {
        tinyIndex += 1;
        tiny2Url.push_back(longUrl);
        url2Tiny[longUrl] = tinyIndex;
        return "http://tinyurl.com/" + to_string(tinyIndex);
    }

    // Decodes a shortened URL to its original URL.
    string decode(const string &shortUrl) {
        int idx = shortUrl.rfind('/') + 1;
        return tiny2Url[stoi(shortUrl.substr(idx))];
    }
};

// Your Solution object will be instantiated and called as such:
// Solution solution;
// solution.decode(solution.encode(url));
